tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
+    from ratbag_emu.protocol.base import MouseData
 
-    id = test_add_device(client, 'generic-hidpp20')
-    sleep(0.1) # Give time for the event nodes to be created
+    id = add_device(client, 'generic-hidpp20')
+    input_nodes = []
 
-    DeviceHandler.devices[id].verbose = False
+    start = time()
+    while len(input_nodes) == 0:
+        input_nodes = client.get('/devices/{}'.format(id)).json()['input_nodes']
+        sleep(0.1)
+        if time() > start + 3:
+            break
 
     # Claim the event nodes
     event_nodes = []
-    for node in DeviceHandler.devices[id].device_nodes:
+    for node in set(input_nodes):
         fd = open(node, 'rb')
         fcntl.fcntl(fd, fcntl.F_SETFL, os.O_NONBLOCK)
-        device = libevdev.Device(fd)
-        try:
-            device.grab()
-        except libevdev.device.DeviceGrabError:
-            with pytest.warns(UserWarning):
-                warnings.warn("Couldn't grab one of the event nodes", UserWarning)
-        event_nodes.append(device)
+        event_nodes.append(libevdev.Device(fd))
+
+    # make this exit on a timeout
+    from threading import RLock
+
+    events = []
+    def collect_events(stop):
+        nonlocal events
+        while not stop.is_set():
+            for node in event_nodes:
+                events += list(node.events())
+
+    import threading
+    stop_event_thread = threading.Event()
+    event_thread = threading.Thread(target=collect_events, args=(stop_event_thread,))
+    event_thread.start()
 
     # Send event
-    data = {
-        'x': 127,
-        'y': 127
-    }
-    response = client.post('/devices/{}/raw_event'.format(id),
-                           data=json.dumps(data),
-                           content_type='application/json')
+    data = [
+        {
+            'start': 0,
+            'end': 500,
+            'action': {
+                'type': 'xy',
+                'x': 5,
+                'y': 5
+            }
+        }
+    ]
+    response = client.post('/devices/{}/event'.format(id),
+                           json=data)
     assert response.status_code == 200
 
-    # Check if we received the event
-    x = y = 0
+    sleep(1)
+    stop_event_thread.set()
+    event_thread.join()
+
     for node in event_nodes:
-        events = node.events()
-        for e in events:
-            if e.matches(libevdev.EV_REL.REL_X):
-                x += e.value
-            elif e.matches(libevdev.EV_REL.REL_Y):
-                y += e.value
         node.fd.close()
-    assert x == data['x']
-    assert y == data['y']
 
+    m = MouseData()
+    for e in events:
+        if e.matches(libevdev.EV_REL.REL_X):
+            m.x += e.value
+        elif e.matches(libevdev.EV_REL.REL_Y):
+            m.y += e.value
+
+    dpi = response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, 'active')).json()
+    assert m.x == int(data[0]['action']['x'] * 0.0393700787 * dpi)
+    assert m.y == int(data[0]['action']['y'] * 0.0393700787 * dpi)
```

Comment:
for this test it would be more sensible to have constants for `X` and `Y`, use them in `data`  and then re-use them here instead of digging through the dict.

Also, where does the `0.0393...` comes from? General rule is that any not immediately obvious number (i.e. anything other than Pi) needs to have a name or a comment.

Reply:
> for this test it would be more sensible to have constants for X and Y, use them in data and then re-use them here instead of digging through the dict.

This is just a simple example, what if I add more stuff? I would need to have constants for everything.

> Also, where does the 0.0393... comes from? General rule is that any not immediately obvious number (i.e. anything other than Pi) needs to have a name or a comment.

```
1 mm = 0.0393700787 inch
```

Reply:
> This is just a simple example, what if I add more stuff? I would need to have constants for everything.

yes, you need to add as much constants as possible. But here, it's more that @whot requested you to have the following pseudo-code:
```python
x = 5
y = 6
data = [
        {
            'start': 0,
            'end': 500,
            'action': {
                'type': 'xy',
                'x': x,
                'y': y
            }
        }
    ]
# inject event, read them
assert m.x == int(x * MM_TO_INCHES * dpi)
assert m.y == int(y * MM_TO_INCHES * dpi)
```

Reply:
I still don't agree but sure.

Reply:
think of it as separation of logic and structure. you want to test that x is 5, so you have a variable for that. Only one place to change it and it all becomes more readable. and sooner or later you likely need to parametrise this anyway - again having a single place to change this is easier.

that's your logic - you want x to be 5 and test for that. The structure is then where you need to use that x, e.g. in this json file and then again in the test. but those shouldn't ever change based on the value of x, so by having a variable you can guarantee that they remain the same structure.

Reply:
> 1 mm = 0.0393700787 inch

huh, never knew that. I think 25.4 mm per inch is common-enough knowledge, so I've only ever written things as `* 25.4` or `/25.4`. Either way a constant or converter function is safer.